rfc/monorepo.md

```diff
@@ -0,0 +1,557 @@
+
+The monorepo will live at `github.com/spinnaker/spinnaker`.  All other repositories will be archived once the transition is completed.
+
+### Primary/Default Branch Name
+
+The primary/default branch of the monorepo will be renamed from `master` to `main`.
+
+### Yes, Deck Too
+
+This design explicitly integrates `deck` and `deck-kayenta` into the monorepo.  Cross-cutting changes do occur between the UI and the backend, and Deck is already a [Lerna](https://lerna.js.org) monorepo within itself - building all of it from its existing Gradle entry point is simple.
+
+Linking the builds of `deck`, and `deck-kayenta` is not something Gradle can do, but we have two good options here:
+
+1. Use the local `deck-kayenta` code during the build process of `deck`, using `yarn`'s workspaces feature
+1. Just integrate `deck-kayenta` as a regular `deck` module, like all the other Lerna subprojects under `packages/*`
+
+### The `spin` CLI
+
+This project is actually the most difficult to integrate into the monorepo, as golang builds tend to be really finicky about their environment.  I think Github Actions should afford us the flexibility to partition that out reasonably, but that requires some investigative work.
+
+There are definitely concrete benefits to including it in the build cycle - it can directly benefit from using adjacent Gate libraries, and it's much more testable when colocated against the APIs it is calling.  Setting up ephemeral CLI testing environments (on every change to Gate, even), is well within GHA's capabilities.
+
+### Halyard
+
+Halyard will also be included in the monorepo.  While it is a tool everyone wants to see replaced, it has no replacement yet.  It consumes many Spinnaker service libraries, and would greatly benefit from composite builds.  
+
+### `spinnaker-monitoring`
+
+The `spinnaker-monitoring` project will not be included in the monorepo.  The project is frozen, and not part of a release build.  The final revision's artifact will continue to be pinned in Halyard and re-tagged as needed for external consumption.
+
+### Integrating Gradle Build Plugins
+
+Gradle composite can also build plugins to be used in the same build, allowing us to also combine [Spinnaker's Gradle build plugin](https://github.com/spinnaker/spinnaker-gradle-project) into the project itself.
+
+This is the only non-Spinnaker-service version that was pinned in Spinnaker Java projects, and including `spinnaker-gradle-project` into the monorepo enables trivial iteration on the Gradle build process.
+
+### Releases and Versioning
+
+Currently, each Java project within Spinnaker maintains its own semver-shaped version designation pinned within various other Spinnaker projects.  This is unnecessary with composite builds, as the project builds as one cohesive unit without any dependency versions at all.
+
+However, containers, libraries, and other artifacts must still be published, and there is a need for a contract with library consumers (such as plugins).
```

Comment:
>we'll still publish individual containers for clouddriver, orca, etc. OR is the intent to have a SINGLE docker with ALL files?

Yes, individual containers just like today.  I don't think there's a use case for a universal container.

> JARs will be published for cloudriver on every merge of resources like orca or fiat due to dependencies

We don't actually need to publish jars outside of numbered/patch releases (or whatever latest/nightly cadence from `main` we choose to implement).  Those are purely for external library consumers.  Gradle builds will produce jars directly from the code in the repository to be shipped with containers, for example - no external Maven round-trip required.  

For `main` library publishing, I would very much recommend that jars get a full publish cycle each time it is done.  This does create some identical published artifacts with different versions, but there are some weird issues with generation of Maven POMs via the `maven-publish` plugin and [gradle module metadata](https://docs.gradle.org/current/userguide/publishing_gradle_module_metadata.html) that are difficult to work around.  

Both of those things will write specific dependency versions into their respective manifests shipped up to a Maven repository.  The problem is that the "version" in there is entirely made-up - it's whatever we want it to be.  Gradle will build everything according to the "version" that CI says it should be, and that will the version written into POMs and gradle module metadata.  If those jars don't actually exist post-publish, pulling any of the platform modules (all Java `-bom` packages e.g. `kork-bom`) will fail because they won't find the dependencies that they think should exist.  

The alternative is a bunch of trying to find the "last" version of every dependency and somehow telling Gradle about that - an approach I would strongly recommend against.  It is far simpler to just publish every jar with every release.

Reply:
You can see an example of what a platform Maven BOM looks like here (latest `orca-bom`):
https://search.maven.org/artifact/io.spinnaker.orca/orca-bom/8.27.1/pom

Within that you can see specific versions for `fiat`, `kork`, and `orca` libraries, all of which must exist or builds against that `orca-bom` version will fail.

Reply:
SO a key thing to publish still jar wise is plugins.  If you're developing a plugin that uses say clouddriver-core, you need it as a test compile only dep in order to build.  NOT having JARs published would make plugin dev more difficult I'd think.

Reply:
I agree, which is why I recommended a full publish for all jars on release branches and whatever cadence for `main` we choose.  Publishing for plugin development is explicitly called out in the proposal.

Reply:
Publish all jars for release branches and main will help Spinnaker plugin developers with their CI process. Glad to see it as plan.